### DISCUSSION TITLE
custom user routes and controllers in v3.0.0?

### DISCUSSION THREAD

**Discussion by hkovacs** [strapi/strapi#6650] [Legacy Topics]
is there a way to create user routes and controllers in api/?

i tried doing so replicating another model, but found that there is no service found for users. only for the provided models.

my use case is:
i have associations, users, and roles. a user is assigned a single association and a single role.

i would like to get all users containing association id abc and role id xyz.

i figure a get request to an endpoint with both the association id and role id which calls a controller that checks all users for both ids.

does this capability exist?

i need to have access to this route as a public user and i dont want to expose /users to the public...

thank you

**Comment by hkovacs** [comment]
so my challenge was, how do i get email address result for a particular user of a given association and specific role:

thanks to some love from the awesome @derrickmehaffy i was able to get this running as follows:

i made a regular route GET `/mymodel/mypath/:association_id`

then in controllers i used this:
following instructions from https://strapi.io/documentation/v3.x/concepts/queries.html#api-reference
```
const results = await strapi.query('user', 'users-permissions').find({ _limit: 1, association_id: 'ABC', role: 'XYZ'});
console.log(results[0].email);
```
it would be nice if it was possible to define fields to return in the result. I wonder if this is just missing from the docs. anyone know?

**Reply by derrickmehaffy** [reply]
> it would be nice if it was possible to define fields to return in the result. I wonder if this is just missing from the docs. anyone know?

Not currently no but there is a light guide on customizing the response: https://strapi.io/documentation/v3.x/guides/custom-data-response.html

But in the future we certainly want to be able to pass along certain fields to return in the response (REST/Internal) similar to how GraphQL supports it. It is being looked into as a feature.

**Reply by hkovacs** [reply]
> Not currently no but there is a light guide on customizing the response: https://strapi.io/documentation/v3.x/guides/custom-data-response.html

great link. ive been wanting to sanitize! TYSM!

**Comment by hkovacs** [comment]
i found this https://strapi.io/documentation/v3.x/concepts/services.html#single-type which instructs to use populate. so to return only the user result instead of the relationship objects do as follows

```
const results = await strapi.query('user', 'users-permissions').find({ _limit: 1, association_id: 'ABC', role: 'XYZ'}, []);
```
notice the `, []` empty array forces return of only the result object. very helpful :)
